Inheritence Refactoring

guidanoli opened the pull request.

guidanoli commented:
Thanks for the valuable feedback, @GCdePaula !

I have simplified the inheritance graph, I hope you can see it clearly in the source code.
To ease visualization, I can later plot the graph using Mermaid here.

Could you expand on why the pattern of view methods that revert if an action cannot be performed might cause troubles with the node? This strategy aimed to simplify the implementation and readability of the contracts. On `eth_call` requests, the node can check whether the call reverted and why (`data`), which it could then decode into the appropriate Solidity error. All this should be done automatically by the bindings library.

pedroargento commented:
I am not sure I like the name EventEmitter. Its not at all related to what the contract does, which is store the deployment block number. The name is related to the offchain contract but it doesn't make a lot of sense in this codebase. Also, every smart contract is a event emitter.

Maybe something like `DeploymentBlock ` or `DeploymentData`?

guidanoli commented:
Every contract _can_ emit an event, but not every contract does.
For those contracts that do emit an event, we provide the deployment block number.
The only reason we provide this info is for off-chain nodes to fast-sync.

guidanoli commented:
> Is the epoch close+finalize design described anywhere?

It is thoroughly described in the documentation of the `EpochManager` interface.

guidanoli commented:
## :deciduous_tree: Inheritance Tree

Here is the inheritance tree in Mermaid:

```mermaid
graph TD
    OutboxImpl --> EpochManagerImpl & ReentrancyGuard
    EpochManagerImpl --> InboxImpl
    InboxImpl --> EventEmitterImpl
    TokenReceiverImpl --> ERC721Holder & ERC1155Holder
    AppImpl["{Dave,Quorum}AppImpl"] --> OutboxImpl & TokenReceiverImpl
```

## :brain: Rationale

- The leaf node is `EventEmitterImpl`, because it doesn't depend on any other contract.
- `InboxImpl` inherits from `EventEmitterImpl` because it emits events.
- `EpochManagerImpl` inherits from `InboxImpl` because it internally stores epoch boundaries in terms of input indices (used by PRT, and to ensure closed epochs are non-empty).
- `OutboxImpl` inherits from `EpochManagerImpl` because it needs to check whether any given post-epoch outputs root has been finalized before (when validating an output). It also inherits from `ReentrancyGuard` to avoid reentrancy attacks during output execution.
- `DaveAppImpl` inherits from `OutboxImpl` (which inherits from `InboxImpl`) because it needs to have access to input Merkle roots.

## :spiral_notepad: Notes

- `EpochManagerImpl` declares an internal virtual `_isPostEpochStateRootValid`, which is implemented differently by `QuorumAppImpl` and `DaveAppImpl`, because it heavily depends on the consensus algorithm used by the application.
- `EpochManagerImpl` also defines several internal functions that are used by both app implementations, such as `_closeEpoch`, `_preFinalize`, `_finalizeEpoch`, `_getInputIndex{InclusiveLower,ExclusiveUpper}Bound`, and `_isFirstNonFinalizedEpochClosed`.

guidanoli commented:
I've renamed `getNumberOf<X>` view functions as `get<X>Count` and added:

- `getClosedEpochCount` view function, to monitor `EpochClosed` events
- `getDeployedAppCount` view function, to monitor `{Dave,Quorum}AppDeployed` events

tuler commented on src/app/concretes/QuorumAppImpl.sol:
```diff
@@ -0,0 +1,154 @@
+        // If not, mark the validator as having already voted in the epoch.
+        {
+            bytes32 bitmap = getAggregatedVoteBitmap(epochIndex);
+            require(!bitmap.getBitAt(validatorId), VoteAlreadyCastForEpoch());
```
is this error necessary?

guidanoli replied:
This check ensures `Vote` is only emitted on new votes.
It may be misleading to have a `Vote` event for a non-change.

guidanoli replied:
Yes, actually this error ensures that validators cannot vote on multiple post-epoch state roots for a given epoch.

pedroargento commented on src/app/interfaces/Inbox.sol:
```diff
@@ -0,0 +1,52 @@
+    function getInputCount() external view returns (uint256);
+
+    /// @notice Get the number of inputs before the current block.
+    function getInputCountBeforeCurrentBlock() external view returns (uint256);
```
A transaction in the same block may alter the count. This function will ignore it if it happens.

guidanoli replied:
This function is used internally to determine epoch boundaries in terms of inputs.

ZzzzHui commented:
> I am not sure I like the name EventEmitter. Its not at all related to what the contract does, which is store the deployment block number. The name is related to the offchain client but it doesn't make a lot of sense in this codebase. Also, every smart contract is a event emitter.
> 
> Maybe something like `DeploymentBlock ` or `DeploymentData`?

I also think `EventEmitter` is quite general. We could consider using what Pedro suggested and I would add `DeploymentInfo` on top of his list

guidanoli commented on src/portals/IPortal.sol:
```diff
@@ -3,13 +3,5 @@
-    /// @return The input box
-    function getInputBox() external view returns (IInputBox);
-}
+interface IPortal {}
```
You're right, this is a vestigial tail from the refactoring I forgot to clean up.

guidanoli replied:
Removed `Iportal` and `Portal`!

guidanoli commented on src/app/interfaces/EpochManager.sol:
```diff
@@ -0,0 +1,186 @@
+interface EpochManager is EventEmitter {
+    /// @notice An epoch has been closed.
+    /// @param epochIndex The index of the epoch
+    /// @param epochFinalizer The contract that makes the epoch reach finality
```
Because Solidity doesn't support generics, like `EpochManager<T>`, where `T` is the type of the epoch finalizer contract. The interface ID of the epoch finalizer contract is provided by the `getEpochFinalizerInterfaceId` view function.
